File: EXAMENES/Feb19/haskell/Kruskal.hs
----------------------------------------------
-- Estructuras de Datos.  2018/19
-- 2º Curso del Grado en Ingeniería [Informática | del Software | de Computadores].
-- Escuela Técnica Superior de Ingeniería en Informática. UMA
--
-- Examen 4 de febrero de 2019
--
-- ALUMNO/NAME:
-- GRADO/STUDIES:
-- NÚM. MÁQUINA/MACHINE NUMBER:
--
----------------------------------------------

module Kruskal(kruskal, kruskals) where

import qualified DataStructures.Dictionary.AVLDictionary as D
import qualified DataStructures.PriorityQueue.LinearPriorityQueue as Q
import DataStructures.Graph.DictionaryWeightedGraph

fromJust :: Maybe (D.Dictionary a w) -> D.Dictionary a w
fromJust Nothing = D.empty
fromJust (Just d) = d

connectedVertex :: (Ord a, Eq a) => WeightedEdge a w -> D.Dictionary a a ->((Bool, a), a)
connectedVertex (WE v w u) dict = (aux v u dict, u)
    where
        aux v u dict = (getRep v dict == getRep u dict, getRep v dict)
            where
                getRep vertex dict
                    | vertex == value = vertex
                    | otherwise = getRep value dict
                        where
                            value = fromJust (D.valueOf vertex dict)


kruskal :: (Ord a, Ord w) => WeightedGraph a w -> [WeightedEdge a w]
kruskal g = aux dict pq []
    where
        pq = foldr Q.enqueue Q.empty (edges g)
        dict = foldl (\res v -> D.insert v v res) D.empty (vertices g)
        aux pq dict ws
            | Q.isEmpty = ws
            | otherwise = if condition then aux Q.dequeue dict ws else aux Q.dequeue (D.insert rep u dict) ([edge]++ws)
                where 
                    edge = Q.first pq
                    ((condition, rep),u) = connectedVertex edge dict


-- Solo para evaluación continua / only for part time students
kruskals :: (Ord a, Ord w) => WeightedGraph a w -> [[WeightedEdge a w]]
kruskals = undefined

File: EXAMENES/Feb19/java/src/Kruskal.java
/**----------------------------------------------
 * -- Estructuras de Datos.  2018/19
 * -- 2º Curso del Grado en Ingeniería [Informática | del Software | de Computadores].
 * -- Escuela Técnica Superior de Ingeniería en Informática. UMA
 * --
 * -- Examen 4 de febrero de 2019
 * --
 * -- ALUMNO/NAME:
 * -- GRADO/STUDIES:
 * -- NÚM. MÁQUINA/MACHINE NUMBER:
 * --
 * ----------------------------------------------
 */

import dataStructures.graph.WeightedGraph;
import dataStructures.graph.WeightedGraph.WeightedEdge;

import dataStructures.dictionary.Dictionary;
import dataStructures.dictionary.HashDictionary;
import dataStructures.priorityQueue.PriorityQueue;
import dataStructures.priorityQueue.LinkedPriorityQueue;
import dataStructures.set.Set;
import dataStructures.set.HashSet;
import dataStructures.tuple.Tuple2;

public class Kruskal {

	private static <V> V getRepresentante(V vertex, Dictionary<V,V> dict){
		V v = vertex;
		while(v!=dict.valueOf(v)){
			v= dict.valueOf(v);
		}
		return v;
	}

	private static <V,W> Tuple2<Tuple2<Boolean,V>,V> areConnected(WeightedEdge<V,W> edge, Dictionary<V,V> dict){
		V repO = getRepresentante(edge.source(), dict);
		V repD = getRepresentante(edge.destination(), dict);
		return new Tuple2<>(new Tuple2<>(repO.equals(repD),repO), edge.destination());
	}
	public static <V,W> Set<WeightedEdge<V,W>> kruskal(WeightedGraph<V,W> g) {

		// COMPLETAR
		Dictionary<V,V> dict = new HashDictionary<>();
		PriorityQueue<WeightedEdge<V,W>> pq = new LinkedPriorityQueue<>();
		Set<WeightedEdge<V,W>> sol = new HashSet<>();

		for(V vertex: g.vertices()){
			dict.insert(vertex,vertex);
		}
		for(WeightedEdge<V,W> edge: g.edges()){
			pq.enqueue(edge);
		}
		while (!pq.isEmpty()){
			WeightedEdge<V,W> edge = pq.first();
			Tuple2<Tuple2<Boolean, V>,V> tupla = areConnected(edge,dict);
			if(!tupla._1()._1()){//if boolean = false
				dict.insert(tupla._1()._2(),tupla._2());
				sol.insert(edge);
			}
			pq.dequeue();
		}
		return sol;
	}

	// Sólo para evaluación continua / only for part time students
	public static <V,W> Set<Set<WeightedEdge<V,W>>> kruskals(WeightedGraph<V,W> g) {

		// COMPLETAR
		
		return null;
	}
}
